snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/converter/SnomedConceptConverter.java

```diff
@@ -307,12 +307,24 @@ private void expandDescendants(List<ISnomedConcept> results, final Set<String> c
 				}
 				
 				final int offset = getOffset(expandOptions);
-				final Collection<String> componentIds = newHashSet(descendantsByAncestor.values());
+				final Set<String> componentIds = newHashSet(descendantsByAncestor.values());
 				
 				if (limit > 0 && !componentIds.isEmpty()) {
+					final Set<String> alreadySeenDescendantIds = newHashSet(Sets.intersection(componentIds, conceptIds));
 					// remove any already known concept definition
 					componentIds.removeAll(conceptIds);
```

Comment:
It would be easier to read, yes, but in certain cases, re-fetching would override an existing instance with a new clean version, so we would have to copy over other expanded properties (like `descriptions`, etc.) to these new instances.